README.md
# Markov Chain and Finite-State Stochastic Machine

This package implements functionality for analyzing stochastic (or random)
finite state (Markov) processes.

It can build a Markov chain from the states of the process.

Also the package provides non-deterministic FSM (finite-state machine)
functionality for evaluating of Markov chains. You can easily build 
a probabilistic automaton from the Markov chain.

One more feature is an integration with amazing [Graphviz](http://www.graphviz.org/) tool.
`markovfsm` plots a state transitions graph of Markov model for you.

# Installation
As usual, `pip install markovfsm`

No special prerequisites required, expect the Graphviz, if you want to use plotting features.
In such case, you have to install Graphviz and [`graphviz` PyPI package](https://pypi.org/project/graphviz/).

# Usage examples
## Coin flipping
Let's start with the simplest Markov process - a flipping of perfect coin.
Let '0' state be 'tails' and '1' state correspond to 'heads'.

We define a random function (`coin()`), once being evaluated, returns the next state of the process.

Let's create a `Chain` object and perform big enough count of experiments:
```python
from random import random
from graphviz import Digraph

from markovfsm import Chain
from markovfsm.plot import transitions_to_graph


def coin():               # random process: the perfect coin flipping
    return 1 if random() > 0.5 else 0

chain = Chain(2, coin())  # create an empty Markov chain with 2 states

for i in range(1000000):  # let the Markov chain build state transition matrix
    chain.learn(coin())   # based on 1000000 of coin flips
```
Let's plot a graph of state transitions.
```python
g = Digraph(format='svg', engine='dot',
            graph_attr={'pad': '0.1', 'nodesep': '0.4', 'ranksep': '1.0'},
            node_attr={'fontname': 'Helvetica'},
            edge_attr={'fontsize': '8.0', 'fontname': 'Helvetica'})

transitions_to_graph(g, chain.transition_matrix(),
                     lambda s: "tails" if s == 0 else "heads")

g.render('./graph')       # this will output ``graph.svg`` (SVG graphics)
                          # and ``graph`` (DOT language) files
```

Example output:
![Coin-flip: state diagram of Markov chain](https://raw.githubusercontent.com/vladignatyev/markovfsm/master/coin.svg?sanitize=true)

## Rigged dice
Another illustrative example is the process of rolling a rigged dice.
We use beta distribution to emulate non-perfect dice.

The remaining part of the code almost the same.
```python
#!coding:utf-8

from random import betavariate
from graphviz import Digraph

from markovfsm import Chain
from markovfsm.plot import transitions_to_graph


def rigged_dice():
    return int((betavariate(0.5, 0.7) * 6))

chain = Chain(6, rigged_dice())

# roll dice many times to build Markov chain for this process
for i in range(100000):
    chain.learn(rigged_dice())
```

And then let's plot states transitions.
```python
def state_mapping(state):
    if state == 0: return u'⚀'
    if state == 1: return u'⚁'
    if state == 2: return u'⚂'
    if state == 3: return u'⚃'
    if state == 4: return u'⚄'
    if state == 5: return u'⚅'

g = Digraph(format='svg', engine='dot',
            graph_attr={'pad': '0.1', 'nodesep': '0.15', 'ranksep': '0.5'},
            edge_attr={'fontsize': '6.0', 'fontname':'Helvetica'})

transitions_to_graph(g, chain.transition_matrix(), state_mapping)
g.render('./graph')
```

Example output:
![Rigged dice: state diagram of Markov chain](https://raw.githubusercontent.com/vladignatyev/markovfsm/master/rigged-dice.svg?sanitize=true)


# Probabilistic finite-state machine
Finite-state machine (FSM, or state machine) is a model of computation.
The machine can be in exactly one of finite number of states.
Probabilistic automaton is a FSM where transitions between states are probabilistic.
Unlike normal FSM, requiring only a graph of possible transitions between states,
probabilistic automaton adds probability of every transition.
```python
from random import random

# build Markov chain with 2 states, init with random state
chain = Chain(2, 0 if random() > 0.5 else 1)

# flip coin many times and build Markov chain for this process
# let 0 be heads and 1 tails
for i in range(1000000):
    chain.learn(0 if random() > 0.5 else 1)

# get transition matrix
#   It should look like:
#
#    P = | 0.5 0.5 |
#        | 0.5 0.5 |
#
P = chain.transition_matrix()

print "%s %s" % (P[0][0], P[0][1])
print "%s %s" % (P[1][0], P[1][1])

# get probabilities of transition from state 0 to other states (0 and 1)
# actually, the line in the transition matrix
print chain.get_transitions_probs(0)

# let's make a FSM with stochastic properties equal to described by Markov chain
# use rnd() as a random numbers generator, and 0 (heads) as initial state
fsm = FSM(chain, 0)

fsm.next()  # will change the state of automaton randomly in a such way that
            # the statistics of such transition will be equal to Markov process
            # statistics
```

# API
`chain.transition_matrix()` will return transition matrix: a matrix N x N,
where N is the number of states, where each i-row correspond to the state of the process
and each j-element in the row contains the probability of transition to state ``j``
from the state ``i``.

`chain.learn(state)` will learn Markov chain from new `state` transition

`FSM(chain, initial_state)` - object, representing probabilistic automaton,
built from `chain` in state `initial_state`


# License
MIT License. Creative Commons CC0.


NEWS.txt
News
====

0.2
---

*Release date: 25-Mar-2019*

* Updated usage information and added illustrations
* Better documentation


0.1
---

*Release date: 20-Dec-2017*

* Initial release


src/markovfsm/plot.py
#!coding:utf-8


def transitions_to_dotlang(matrix):
    dim = len(matrix)
    transitions = []
    for current_state in range(dim):
        for next_state in range(dim):
            t = '   %s -- %s [label = "%.4f"]' % (current_state, next_state, matrix[current_state][next_state])
            transitions.append(t)

    return "digraph G {\n%s\n}" % ('\n'.join(transitions))


def transitions_to_graph(digraph, matrix, state_index_to_name = lambda s: "%s" % s):
    dim = len(matrix)
    for i in range(dim):
        digraph.node(state_index_to_name(i))
    for current_state in range(dim):
        for next_state in range(dim):
            alpha = "{0:x}".format(int(round(matrix[current_state][next_state] * 250 + 6)))
            digraph.edge(state_index_to_name(current_state),
                         state_index_to_name(next_state),
                         label=" %.4f " % matrix[current_state][next_state],
                         color='#ff0000%s' % alpha,
                         fontcolor='#f00000%s' % alpha)

if __name__ == '__main__':
    from graphviz import Digraph

    matrix = [[0.5, 0.5], [0.5, 0.5]]
    print transitions_to_dotlang(matrix)
    g = Digraph(format='svg', engine='dot',
                graph_attr={'overlap': 'true', 'pad': '0.1', 'nodesep': '0.5', 'ranksep': '1.0'},
                edge_attr={'fontname':'HelveticaNeue'})

    transitions_to_graph(g, matrix, lambda s: "heads" if s == 1 else "tails")
    g.render('./graph')


examples/double_dice.py
#!coding:utf-8

from random import random
from graphviz import Digraph

from markov import Chain
from markov.plot import transitions_to_graph


def double_dice():
    return int(random() * 6) + int(random() * 6)

chain = Chain(11, double_dice())

# flip coin many times and build Markov chain for this process
# let 0 be heads and 1 tails
for i in range(1000000):
    chain.learn(double_dice())


g = Digraph(format='svg', engine='dot',
            graph_attr={'overlap': 'false', 'pad': '0.1', 'nodesep': '0.35', 'ranksep': '0.35'},
            edge_attr={'fontname':'Helvetica', 'fontsize': '8.0'},
            node_attr={'fontname': 'Helvetica'})

def state_mapping(s):
    scores = {
        2: u'⚀⚀',
        3: u'⚀⚁',
        4: u'⚁⚁',
        5: u'⚁⚂',
        6: u'⚂⚂',
        7: u'⚂⚃',
        8: u'⚃⚃',
        9: u'⚃⚄',
        10: u'⚄⚄',
        11: u'⚄⚅',
        12: u'⚅⚅'
    }
    return scores[s + 2]

transitions_to_graph(g, chain.transition_matrix(), state_mapping)
g.render('./graph')


examples/rigged_dice.py
#!coding:utf-8

from random import betavariate
from graphviz import Digraph

from markov import Chain
from markov.plot import transitions_to_graph


def rigged_dice():
    return int((betavariate(0.5, 0.7) * 6))

chain = Chain(6, rigged_dice())

# roll dice many times and build Markov chain for this process
for i in range(100000):
    chain.learn(rigged_dice())


def state_mapping(state):
    if state == 0: return u'⚀'
    if state == 1: return u'⚁'
    if state == 2: return u'⚂'
    if state == 3: return u'⚃'
    if state == 4: return u'⚄'
    if state == 5: return u'⚅'



g = Digraph(format='svg', engine='dot',
            graph_attr={'pad': '0.1', 'nodesep': '0.15', 'ranksep': '0.5'},
            edge_attr={'fontsize': '6.0', 'fontname':'Helvetica'})

transitions_to_graph(g, chain.transition_matrix(), state_mapping)
g.render('./graph')


src/markovfsm/__init__.py
from copy import copy
from random import random


class Chain(object):
    def __init__(self, states_count, initial_state):
        self._states_count = states_count
        self._initial_state = initial_state
        self.clear()

    """
        Build Markov chain from data by feeding new state of the random process
    """
    def learn(self, state):
        self.states[self.current_state][state] = self.states[self.current_state][state] + 1
        self.current_state = state

    """
        Get the transition matrix P of Markov chain
    """
    def transition_matrix(self):
        m = copy(self.states)
        for i in range(self._states_count):
            s = 1.0 / sum([m[i][j] for j in range(self._states_count)])
            m[i] = [m[i][j] * s for j in range(self._states_count)]
        return m

    """
        Get probabilities of transitions to another states from current state
    """
    def get_transitions_probs(self, state):
        s = 1.0 / sum([self.states[state][j] for j in range(self._states_count)])
        c = copy(self.states[state])
        return [c[i] * s for i in range(self._states_count)]

    """
        Clear the model
    """
    def clear(self):
        self.states = [[0 for j in range(self._states_count)] for i in range(self._states_count)]
        self.current_state = self._initial_state



class FSM(object):
    def __init__(self, chain, initial_state, rnd=random):
        self.chain = chain
        self.current_state = initial_state
        self._rnd = rnd

    @property
    def state(self):
        return self.current_state

    def next(self):
        probs = self.chain.get_transitions_probs(self.current_state)
        r = self._rnd()

        p0 = 0.0
        p1 = 0.0

        for i in range(len(probs)):
            p1 = p0 + probs[i]
            if r >= p0 and r < p1:
                self.current_state = i
                return self.current_state
            else:
                p0 = p1



if __name__ == '__main__':
    from random import random

    # build Markov chain with 2 states, init with random state
    chain = Chain(2, 0 if random() > 0.5 else 1)

    # flip coin many times and build Markov chain for this process
    # let 0 be heads and 1 tails
    for i in range(1000000):
        chain.learn(0 if random() > 0.5 else 1)

    # get transition matrix
    #   It should look like:
    #
    #    P = | 0.5 0.5 |
    #        | 0.5 0.5 |
    #
    P = chain.transition_matrix()

    print "%s %s" % (P[0][0], P[0][1])
    print "%s %s" % (P[1][0], P[1][1])

    # get probabilities of transition from state 0 to other states (0 and 1)
    # actually, the line in the transition matrix
    print chain.get_transitions_probs(0)

    # let's make a FSM with stochastic properties equal to described by Markov chain
    # use rnd() as a random numbers generator, and 0 (heads) as initial state
    fsm = FSM(chain, 0)

    # let's make another chain, build upon random process emulated by FSM
    chain2 = Chain(2, 0)

    # make a chain with states produced by FSM
    for j in range(1000000):
        chain2.learn(fsm.next())

    # get transition matrix. it should be nearly equal
    P2 = chain2.transition_matrix()

    print "%s %s" % (P2[0][0], P2[0][1])
    print "%s %s" % (P2[1][0], P2[1][1])

    print chain2.get_transitions_probs(0)


examples/coin.py
#!coding:utf-8

from random import random
from graphviz import Digraph

from markov import Chain
from markov.plot import transitions_to_graph


def coin():               # random process: the perfect coin flipping
  return 1 if random() > 0.5 else 0

chain = Chain(2, coin())  # create an empty Markov chain with 2 states

for i in range(1000000):  # let the Markov chain build state transition matrix
  chain.learn(coin())   # based on 1000000 of coin flips

# let's plot a graph of state transitions
g = Digraph(format='svg', engine='dot',
          graph_attr={'pad': '0.1', 'nodesep': '0.4', 'ranksep': '1.0'},
          node_attr={'fontname': 'Helvetica'},
          edge_attr={'fontsize': '8.0', 'fontname': 'Helvetica'})

# the lambda here is a simple mapping function to show beautiful name
# instead of weird '0' or '1'

transitions_to_graph(g, chain.transition_matrix(),
                   lambda s: "tails" if s == 0 else "heads")

g.render('./graph')       # this will output ``graph.svg`` (SVG graphics)
                        # and ``graph`` (DOT language) files
